server/src/main/java/org/opensearch/rest/NamedRoute.java

```diff
@@ -9,47 +9,168 @@
+        private String path;
+        private String uniqueName;
+        private final Set<String> legacyActionNames = new HashSet<>();
+        private Function<RestRequest, RestResponse> handler;
+
+        /**
+         * Sets the REST method for the route.
+         *
+         * @param method the REST method for the route
+         * @return the builder instance
+         */
+        public Builder method(RestRequest.Method method) {
+            this.method = method;
+            return this;
+        }
+
+        /**
+         * Sets the URL path for the route.
+         *
+         * @param path the URL path for the route
+         * @return the builder instance
+         */
+        public Builder path(String path) {
+            this.path = path;
+            return this;
+        }
+
+        /**
+         * Sets the name for the route.
+         *
+         * @param name the name for the route
+         * @return the builder instance
+         */
+        public Builder uniqueName(String name) {
+            this.uniqueName = name;
+            return this;
+        }
+
+        /**
+         * Sets the legacy action names for the route.
+         *
+         * @param legacyActionNames the legacy action names for the route
+         * @return the builder instance
+         */
+        public Builder legacyActionNames(Set<String> legacyActionNames) {
+            this.legacyActionNames.addAll(validateLegacyActionNames(legacyActionNames));
+            return this;
+        }
+
+        /**
+         * Sets the handler for this route
+         *
+         * @param handler the handler for this route
+         * @return the builder instance
+         */
+        public Builder handler(Function<RestRequest, RestResponse> handler) {
+            if (handler == null) {
+                throw new OpenSearchException("Route handler must not be null.");
+            }
+            this.handler = handler;
+            return this;
         }
-        return routeName.matches(VALID_ACTION_NAME_PATTERN);
+
+        /**
+         * Builds a new instance of {@link NamedRoute} based on the provided parameters.
+         *
+         * @return a new instance of {@link NamedRoute}
+         * @throws OpenSearchException if the route name is invalid
+         */
+        public NamedRoute build() {
+            return new NamedRoute(this);
+        }
+
+        private Set<String> validateLegacyActionNames(Set<String> legacyActionNames) {
+            if (legacyActionNames == null) {
+                return new HashSet<>();
+            }
+            for (String actionName : legacyActionNames) {
+                if (!TransportService.isValidActionName(actionName)) {
+                    throw new OpenSearchException(
+                        "Invalid action name [" + actionName + "]. It must start with one of: " + TransportService.VALID_ACTION_PREFIXES
+                    );
+                }
+            }
+            return legacyActionNames;
```

Comment:
Usually `validate()` methods return `void` and just throw exceptions.  Looks like the only usefulness of this returning something is just replacing null with an empty set (prefer `Collections.emptySet()` for that) but I think it's fine to disallow null to begin with.

Reply:
agreed. I'll change the method call.